architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+### Architecture Diagram
+
+Resource type registration and querying is implemented in UCP. Any client may call the registration APIs or query APIs including resource providers.
+
+<center>
+
+![architecture diagram](./2024-08-user-defined-types-registration/registration.png)
+
+**Fig: Diagram of resource type registration and querying** [link](https://excalidraw.com/#json=FzQQaxv1d-vlmmDBAeOfc,UUXIasjO8p27hN0AB8qeiw)
+
+</center>
+
+### Detailed Design
+
+Since this is our first nested resource type, a basic primer on nested types is needed. A nested resource type exists *inside* another resource type. A resource type is that is not nested is *top-level*. These relationships are also sometimes referred to as a parent and child. 
+
+The following statements are all accurate:
+
+- The `System.Resources/resourceProviders` type is the parent of `System.Resources/resourceProviders/resourceTypes`. 
+- The `System.Resources/resourceProviders/resourceTypes` type is the child of `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders/resourceTypes` type is nested within `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders` type is a top-level resource type.
+
+Any resource type with two segments (separated by `/`) is a top-level resource type. Any resource type with 3 or more segments (separate by `/`) is a nested resource type.
+
+The resource ids (and URLs) are written like the following:
+
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders` - List all resource providers.
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example` - Get the Contoso.Example resource provider.
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example/resourceTypes` - List all resource types inside the `Contoso.Example` resource provider/namespace.
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example/resourceTypes/contosoBuses` - Gets the `contosoBuses` resource type inside the `Contoso.Example` resource provider/namespace.
+
+Nesting a resource type also nests the lifecyle of the resource. Using our examples above, the `/planes/radius/{plane}/providers/System.Resources/resourceProviders/resourceTypes/Contoso.Example/contosoBuses` type cannot exist unless `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example` already exists.
+
+There are a few simple rules for this:
+
+- For operations on parent resources...
+  - A `DELETE` operation on a parent resource deletes all children.
+
+- For operations on child resources...
+  - A *CRUL* operation (anything but delete) on a child resource must return a `404` when the parent does not exist.
+  - A `DELETE` operation on a child resource never returns a `404`.
```

Comment:
Unless someone has already deleted that maybe?

Reply:
`DELETE `operations in UCP never return a 404. `DELETE` is the "make it gone" button.